pulsar-client/src/test/java/org/apache/pulsar/client/impl/MessageIdAdvUtilsTest.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import static org.testng.Assert.assertEquals;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.BitSet;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.Cleanup;
+import org.apache.commons.lang3.RandomUtils;
+import org.apache.pulsar.client.api.MessageIdAdv;
+import org.testng.annotations.Test;
+
+/**
+ * Unit test for {@link MessageIdAdvUtils}.
+ */
+public class MessageIdAdvUtilsTest {
+
+    /**
+     * Call <code>acknowledge</code> concurrently with batch message, and verify that only return true once
+     *
+     * @see MessageIdAdvUtils#acknowledge(MessageIdAdv, boolean)
+     * @see MessageIdAdv#getAckSet()
+     */
+    @Test
+    public void testAcknowledgeIndividualConcurrently() throws InterruptedException {
+        ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("pulsar-consumer-%d").build();
+        @Cleanup("shutdown")
+        ExecutorService executorService = Executors.newFixedThreadPool(16, threadFactory);
```

Comment:
the thread pool with a fixed size of 16 seems like a problem if there's a need to have more concurrency than 16 threads. I guess using `newCachedThreadPool` would be suitable for a test to let the thread pool grow to the required size.

Reply:
Sounds good, changed to `newCachedThreadPool`